examples/move/closed_loop/sources/closed_loop.move

```diff
@@ -0,0 +1,447 @@
+    struct TokenPolicyCap<phantom T> has key, store { id: UID, for: ID }
+
+    /// A single object representing the Policy the Token.
+    ///
+    /// Policy defines what actions can be performed on the token, and what
+    /// rules must be satisfied for the action to go through.
+    struct TokenPolicy<phantom T> has key {
+        id: UID,
+        /// The balance that is effectively burned by the user (and the user
+        /// claims a storage rebate), however actual supply deduction is only
+        /// performed by Admin when necessary.
+        ///
+        /// This balance can never be withdrawn by anyone and only burned by
+        /// the Admin.
+        burned_balance: Balance<T>,
+        /// The set of rules that define what actions can be performed on the
+        /// token. Each rule contains the set of `TypeName`s that must be
+        /// received by the `ActionRequest` for the action to be performed.
+        rules: VecMap<String, vector<TypeName>>
+    }
+
+    /// A request to perform an "Action" on a token.
+    struct ActionRequest<phantom T> {
+        /// Name of the Action to look up in the Policy.
+        ///
+        /// > String will always be shorter than the fully qualfied name of the
+        /// type (just in case the gas micro optimizations matter).
+        name: String,
+        /// Amount is present in all of the txs
+        amount: u64,
+        /// Sender is a permanent field always
+        sender: address,
+        /// Recipient is only available in transfer operation
+        recipient: Option<address>,
+        /// Collected approvals from Rules.
+        approvals: VecSet<TypeName>,
+    }
+
+    /// Create a new `TokenPolicy` and a matching `TokenPolicyCap`.
+    /// The `TokenPolicy` must then be shared using `share_policy`.
+    public fun new<T>(
+        _treasury_cap: &mut TreasuryCap<T>, ctx: &mut TxContext
+    ): (TokenPolicy<T>, TokenPolicyCap<T>) {
+        let policy = TokenPolicy {
+            id: object::new(ctx),
+            burned_balance: balance::zero(),
+            rules: vec_map::empty()
+        };
+
+        let cap = TokenPolicyCap {
+            id: object::new(ctx),
+            for: object::id(&policy)
+        };
+
+        (policy, cap) // we don't share it to use in `init()`
+    }
+
+    // === Token Actions ===
+
+    /// Transfer a `Token` to a `recipient`.
+    public fun transfer<T>(
+        t: Token<T>, recipient: address, ctx: &mut TxContext
+    ): ActionRequest<T> {
+        let amount = balance::value(&t.balance);
+        transfer::transfer(t, recipient);
+
+        new_request(string::utf8(TRANSFER), amount, option::some(recipient), ctx)
+    }
+
+    /// Spend a `Token` by converting it into a `SpentToken`.
+    public fun spend<T>(t: Token<T>, ctx: &mut TxContext): ActionRequest<T> {
+        let Token { id, balance } = t;
+        let amount = balance::value(&balance);
+
+        object::delete(id);
+        transfer::transfer(
```

Comment:
If `spend` took `&mut TokenPolicy` as an argument, we could dump `balance` directly into `burned_balance`. I think this would let us eliminate `SpentToken` + the extra user effort of burning `SpentToken` to claim the rebate.

Thoughts on this simplification?

Reply:
My initial thought was that we'd need to resolve the request in the purchase function. But then yeah, you won't be able to confirm and have to take the request, pass all the rules and bring it back to the `TokenPolicy`. This way it would work.

```move
module 0x0::spend {
  fun purchase(self: &mut TokenPolicy<MEOW>, token: Token<MEOW>, /* ctx */ ): ActionRequest<MEOW> {
    let request = closed_loop::spend(self, token, ctx);
    request
  }
}
```

Another idea (would love your thoughts on this) - extend `ActionRequest` to include `Option<Balance>` in it, so that we can allow for certain actions to "burn" Balance. We already have this pattern for `Option<receiver>`, and it could be a nice extension for the base which gives even more flexibility for arbitrary rules.

Reply:
I like the `Option<Balance>` idea--easy to imagine that being useful in future request types.

Reply:
It comes with strings attached - we now require mutability on Policy always.